pkg/plugins/policies/meshaccesslog/api/v1alpha1/validator.go

```diff
@@ -0,0 +1,150 @@
+			for idx, field := range format.GetJson() {
+				indexedField := backendIndexed.Field("json").Index(idx)
+				if field.GetKey() == "" {
+					verr.AddViolationAt(indexedField.Field("key"), `key cannot be empty`)
+				}
+				if field.GetValue() == "" {
+					verr.AddViolationAt(indexedField.Field("value"), `value cannot be empty`)
+				}
+			}
+		}
+	}
+}
+
+func (r *MeshAccessLogResource) validateToOrFromDefined(spec validators.PathBuilder, verr *validators.ValidationError) {
+	if len(r.Spec.GetFrom()) == 0 && len(r.Spec.GetTo()) == 0 {
+		verr.AddViolationAt(spec, `at lest one of "from", "to" has to be defined`)
+	}
+}
+
+func (r *MeshAccessLogResource) validateIncompatibleCombinations(spec validators.PathBuilder, verr *validators.ValidationError) {
+	to := spec.Field("to")
+	targetRef := r.Spec.GetTargetRef().GetKindEnum()
+	if targetRef == common_proto.TargetRef_MeshGatewayRoute && len(r.Spec.GetTo()) > 0 {
+		verr.AddViolationAt(to, `cannot use "to" when "targetRef" is "MeshGatewayRoute" - there is no outbound`)
+	}
+	if targetRef == common_proto.TargetRef_MeshHTTPRoute && len(r.Spec.GetTo()) > 0 {
+		verr.AddViolationAt(to, `cannot use "to" when "targetRef" is "MeshHTTPRoute" - "to" always goes to the application`)
+	}
+}
+
+func (r *MeshAccessLogResource) validateTo(spec validators.PathBuilder, verr *validators.ValidationError) {
+	to := spec.Field("to")
+	for idx, toItem := range r.Spec.GetTo() {
+		targetRefErr := matcher_validators.ValidateTargetRef(to.Index(idx).Field("targetRef"), toItem.GetTargetRef(), &matcher_validators.ValidateTargetRefOpts{
+			SupportedKinds: []common_proto.TargetRef_Kind{
+				common_proto.TargetRef_Mesh,
+				common_proto.TargetRef_MeshService,
+			},
+		})
+		verr.AddError("", targetRefErr)
+
+		if toItem.GetDefault() == nil {
+			verr.AddViolationAt(to.Index(idx).Field("default"), "must be defined")
+		}
+
+		toIndexed := to.Index(idx)
+		for backendIdx, backend := range toItem.GetDefault().GetBackends() {
+			backendIndexed := toIndexed.Field("default").Field("backend").Index(backendIdx)
+			r.validateBackend(backend, verr, backendIndexed)
+		}
+	}
+}
+
+func (r *MeshAccessLogResource) validateFrom(spec validators.PathBuilder, verr *validators.ValidationError) {
+	from := spec.Field("from")
+	for idx, fromItem := range r.Spec.GetFrom() {
+		targetRefErr := matcher_validators.ValidateTargetRef(from.Index(idx).Field("targetRef"), fromItem.GetTargetRef(), &matcher_validators.ValidateTargetRefOpts{
+			SupportedKinds: []common_proto.TargetRef_Kind{
+				common_proto.TargetRef_Mesh,
+			},
+		})
+		verr.AddError("", targetRefErr)
+
+		if fromItem.GetDefault() == nil {
+			verr.AddViolationAt(from.Index(idx).Field("default"), "must be defined")
+		}
+
+		toIndexed := from.Index(idx)
+		for backendIdx, backend := range fromItem.GetDefault().GetBackends() {
+			backendIndexed := toIndexed.Field("default").Field("backend").Index(backendIdx)
+			r.validateBackend(backend, verr, backendIndexed)
```

Comment:
Using `verr. AddViolationAt()` wouldn't be simpler than passing `backendIndexed` down?

Reply:
I'm using `verr.AddViolationAt()` in `validateBackend()`. Do you want me to pass a version (higher order function) of `verr.AddViolationAt()` with fixed first parameter? Something like this:

```
func addViolationAtPath(verr *validators.ValidationError, path validators.PathBuilder) func(message string) {
	return func(message string) {
		verr.AddViolationAt(path, message)
	}
}
```

I could do that but then in [validateBackend](https://github.com/kumahq/kuma/pull/4998/files#diff-514fa6f8fc6be662c82f1dcb020fbf431c014218a289efe0645deae0664f5b2eR34) I need to construct a "deeper" path to correctly mark the error with `path` not being present.

Reply:
We need `backendIndexed` to build other paths, we could return `verr` instead of passing it everywhere and then merge it back up but is seems like it would just be extra work that I don't think adds too much value (those functions are not "pure" but the exposed `validate()` function is).